src/app/birthday/birthday.component.spec.ts

```diff
@@ -1,17 +1,87 @@
 /* tslint:disable:no-unused-variable */
-import { TestBed } from '@angular/core/testing';
+import {async, ComponentFixture, fakeAsync, TestBed, tick} from '@angular/core/testing';
 
-import { BirthdayComponent } from './birthday.component';
+import {BirthdayComponent} from './birthday.component';
+import {BirthdayService} from './birthday.service';
+import {CUSTOM_ELEMENTS_SCHEMA, SimpleChange} from '@angular/core';
+import {Subject} from 'rxjs';
+import {By} from '@angular/platform-browser';
+
+class BirthServiceMock {
+  getBirthdaySubject$ = new Subject();
+  getBirthdayFor = jasmine.createSpy().and.returnValue(this.getBirthdaySubject$);
+}
 
 describe('BirthdayComponent', () => {
-  beforeEach(() => {
+  beforeEach(async(() => {
     TestBed.configureTestingModule({
       declarations: [BirthdayComponent],
+      providers: [{provide: BirthdayService, useClass: BirthServiceMock}],
+      schemas: [CUSTOM_ELEMENTS_SCHEMA],
     }).compileComponents();
+  }));
+
+  let fixture: ComponentFixture<BirthdayComponent>;
+  const name = 'Anya';
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(BirthdayComponent);
+    fixture.componentInstance.name = name;
+    fixture.componentInstance.ngOnChanges({
+      name: new SimpleChange(null, fixture.componentInstance.name, true)
+    });
+  });

```

Comment:
Offloading repetitive initialization code to `beforeEach` block is usually a good practice however it's better to not put too much init code there because it takes away freedom to customize specific test cases with custom init logic and also make the test cases (`it` blocks) less readable/understandable because big part of it is extracted somewhere else.

That is why it is important to find a sweet spot for shared init code and one rule that may help in guiding this is to think about which pieces of init logic affecting the test case directly and which are not directly relevant.
For example, if you are testing the component input - then it is probably not a good idea to move input initialization into `beforeEach` as it it direct precondition of the these test cases.

Reply:
Completely agree with you points, however In this particular case all of the tests require initialized input, and I put this logic into before each block to prevent duplication of these logic into each test.